transport/internal/tls/dialer/dialer_test.go

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2022 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dialer
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc/transport/internal/tls/testscenario"
+)
+
+func TestDialer(t *testing.T) {
+	tests := []struct {
+		desc             string
+		withCustomDialer bool
+		data             string
+	}{
+		{desc: "without_custom_dialer", data: "test_no_dialer"},
+		{desc: "with_custom_dialer", data: "test_with_dialer", withCustomDialer: true},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			serverTLSConfig, clientTLSConfig := tlsConfigs(t)
+			params := Params{
+				Config: clientTLSConfig,
+			}
+			calledCustomDialer := false
+			if tt.withCustomDialer {
+				params.Dialer = func(ctx context.Context, network, address string) (net.Conn, error) {
+					calledCustomDialer = true
+					return (&net.Dialer{}).DialContext(ctx, network, address)
+				}
+			}
+			dialer := NewTLSDialer(params)
```

Comment:
can you move this line before it is actually used? L80. It got me confused a bit in the first pass that they we are setting a dialer before we listen.